File: app/src/at-hoangtran/java/com/asiantech/intern20summer1/intro/RainsOfReason.java
package com.asiantech.intern20summer1.intro;

public class RainsOfReason {
    public static void main(String[] args) {
        int replace = 4;
        int substiution = 5;
        String name = "abcxzsd/";
        String cell1 = "A5";
        String cell2 = "B4";
        int[] intArray = new int[]{1, 4, 2, 3, 4, 8};
        System.out.println(arrayReplace(intArray, replace, substiution));
        System.out.println(evenDigitsOnly(replace));
        System.out.println(variableName(name));
        System.out.println(alphabeticShift(name));
        System.out.println(chessBoardCellColor(cell1, cell2));
    }

    static int[] arrayReplace(int[] inputArray, int elemToReplace, int substitutionElem) {
//        Given an array of integers, replace all the occurrences of elemToReplace with substitutionElem.
        for (int i = 0; i < inputArray.length; i++) {
            if (inputArray[i] == elemToReplace) {
                inputArray[i] = substitutionElem;
            }
        }
        return inputArray;
    }

    static boolean evenDigitsOnly(int n) {
//        Check if all digits of the given integer are even.
        String s = Integer.toString(n);
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) % 2 != 0) {
                return false;
            }
        }
        return true;
    }

    static boolean variableName(String name) {
//        Correct variable names consist only of English letters, digits and underscores and they can
//        't start with a digit.
//
//        Check if the given string is a correct variable name.
        return (name.matches("[a-zA-Z_]+[\\w_]*"));
    }

    static String alphabeticShift(String inputString) {
//        Given a string, your task is to replace each of its characters by the next one in the
//        English alphabet;
//        i.e.replace a with b, replace b with c, etc(z would be replaced by a).
        char[] c = new char[inputString.length()];
        for (int i = 0; i < inputString.length(); i++) {
            c[i] = inputString.charAt(i);
            if (c[i] == 'z') {
                c[i] = 'a';
            } else {
                c[i]++;
            }
        }
        return String.copyValueOf(c);
    }

    static boolean chessBoardCellColor(String cell1, String cell2) {
//        Given two cells on the standard chess board, determine whether they have the same color or not.
        return (Math.abs((cell1.charAt(0) - cell2.charAt(0)) % 2)
                == Math.abs((cell1.charAt(1) - cell2.charAt(1)) % 2));
    }
}

File: app/src/at-huybui/java/com/asiantech/intern20summer1/introKotlin/RainsofReason.kt
package com.asiantech.intern20summer1.introKotlin

class RainsofReason {


    /** 25 **
     *Given an array of integers, replace all the occurrences of elemToReplace with substitutionElem.
     */
    fun arrayReplace(
        inputArray: MutableList<Int>,
        elemToReplace: Int,
        substitutionElem: Int
    ): MutableList<Int> {
        return inputArray.map { i -> if (i == elemToReplace) substitutionElem else i }
            .toMutableList()
    }

    /** 26 **
     * Check if all digits of the given integer are even.
     */
    fun evenDigitsOnly(n: Int): Boolean {
        n.toString().forEach { it ->
            if (it.toInt() % 2 == 1) {
                return false
            }
        }
        return true
    }

    /** 27 **
     * Correct variable names consist only of English letters,
     * digits and underscores and they can't start with a digit.
     * Check if the given string is a correct variable name.
     */
    fun variableName(name: String): Boolean {
        val regex = """[a-z|_][\w|_]*""".toRegex() // regex của name
        return regex.matches(name)
    }

    /** 28 **
     *Given a string, your task is to replace each of its characters by the next one in the English alphabet;
     * i.e. replace a with b, replace b with c, etc (z would be replaced by a).
     */
    fun alphabeticShift(inputString: String): String {
        var mOutputString = ""
        inputString.forEach {
            when {
                it == 'z' -> mOutputString += "a"
                else -> mOutputString += it + 1
            }
        }
        return mOutputString
    }

    /** 29 **
     * Given two cells on the standard chess board,
     * determine whether they have the same color or not.
     */
    fun chessBoardCellColor(cell1: String, cell2: String): Boolean {
        return cell1.map { char -> char.toByte() }.sum() % 2 ==
                cell1.map { char -> char.toByte() }.sum() % 2
    }
}
